docs/architecture/adr-042-state-sync.md

```diff
@@ -0,0 +1,209 @@
+# ADR 042: State Sync Design
+
+## Changelog
+
+2019-06-27: Init by EB
+2019-07-04: Follow up by brapse
+
+## Context
+StateSync is a feature which would allow a new node to receive a
+snapshot of the application state without downloading blocks or going
+through consensus. Once downloaded, the node could switch to FastSync
+and eventually participate in consensus. The goal of StateSync is to
+facilitate setting up a new node as quickly as possible.
+
+## Considerations
+Because Tendermint doesn't know anything about the application state,
+the goal of StateSync of to broker messages between nodes and through
+the ABCI. The implementation will have multiple touch points through the
+codebase:
+
+* A StateSync reactor to facilitate peer communication
+* A Set of ABCI messages to transmit application state to the reactor
+* A Set of MultiStore APIs for exposing snapshot data to the ABCI
+* A Storage format with validation and performance conciderations
+
+### Implementation Properties
+Beyond the approach, any implementation of StateSync can be evaluated
+across different criteria:
+
+* Speed: Expected throughput of producing and consuming snapshots
+* Safety: Cost of pushing partial or completely invalid snapshots to a node
+* Effort: How much effort does an implementation require
+
+### Implementation Question
+
+* What is the format of a snapshot
+    * Complete snapshot 
+    * Ordered IAVL key ranges
+    * Compressed individually chunks which can be validated
+* How is data validated
+    * Trust a peer with it's data blindly
+    * Trust a majority of peers
+    * Use light client validation to validate any key based on consensus
+      produced merkle tree root
+* What are the performance characteristics
+    * Random vs sequential reads
+    * How parallelizeable is the scheduling algorithm
+
+### Proposals
+Broadly speaking there are two approaches to this problem which have had
+varying degrees of discussion and progress. These approach can be
+summarized as:
+
+* Lazy: Where snapshots are produced dynamically at request time
```

Comment:
Might be helpful to clarify that `Lazy` leans towards using the native interface of the tree, ie. key-value pairs, where as `Eager` leans towards constructing something more specifically for state syncing purposes, ie. chunks.

Reply:
Agree with Bucky here. It's probably not lazy vs eager, but generating snapshot (potentially different data structure) vs serving requests from original data store. And then in the former case we can differentiate between lazy and eager, where lazy would generate snapshot only when asked.